src/daemon/daemon.cpp

```diff
@@ -2668,6 +2680,84 @@ catch (const std::exception& e)
     status_promise->set_value(grpc::Status(grpc::StatusCode::INTERNAL, e.what(), ""));
 }
 
+void mp::Daemon::clone(const CloneRequest* request,
+                       grpc::ServerReaderWriterInterface<CloneReply, CloneRequest>* server,
+                       std::promise<grpc::Status>* status_promise)
+try
+{
+    config->factory->require_clone_support();
+    mpl::ClientLogger<CloneReply, CloneRequest> logger{mpl::level_from(request->verbosity_level()),
+                                                       *config->logger,
+                                                       server};
+
+    const auto& source_name = request->source_name();
+    const auto [instance_trail, status] = find_instance_and_react(operative_instances,
+                                                                  deleted_instances,
+                                                                  source_name,
+                                                                  require_operative_instances_reaction);
+    if (status.ok())
+    {
+        assert(instance_trail.index() == 0);
+        const auto source_vm_ptr = std::get<0>(instance_trail)->second;
+        assert(source_vm_ptr);
+        const VirtualMachine::State source_vm_state = source_vm_ptr->current_state();
+        if (source_vm_state != VirtualMachine::State::stopped && source_vm_state != VirtualMachine::State::off)
+        {
+            return status_promise->set_value(
+                grpc::Status{grpc::FAILED_PRECONDITION,
```

Comment:
In `daemon::delete` and `daemon::stop` we already use `grpc::FAILED_PRECONDITION` for invalid state (by catching `VMStateInvalidException`). It can not be `grpc::StatusCode::INVALID_ARGUMENT` in those cases because `grpc::StatusCode::INVALID_ARGUMENT` is used in `find_instance_and_react` for the deleted instance or instance not found cases and they have to be distinguishable in order to make cli append suggestive message (using force stop or use purge delete). 

In the case of `daemon::clone` case, it should not make a difference because our cli side does not do anything specific to `grpc::StatusCode::INVALID_ARGUMENT` or `grpc::FAILED_PRECONDITION`. However, for the sake of consistency, it is nice to keep the invalid state to `grpc::StatusCode::FAILED_PRECONDITION`.

Reply:
As discussed, let's use your option and apply to other state based precondition failures.